src/meta/src/hummock/compaction/picker/min_overlap_compaction_picker.rs

```diff
@@ -222,127 +221,105 @@ impl NonOverlapSubLevelPicker {
+                for other in target_tables
+                    .iter()
+                    .take(overlap_files_range.end)
+                    .skip(overlap_files_range.start)
+                {
                     if level_handler.is_pending_compact(&other.sst_id) {
                         pending_compact = true;
                         break;
                     }
-                    debug_assert!(!select_sst_id_set.contains(&other.sst_id));
-                    add_files_size += other.file_size;
-                    overlap_info.update(other);
-                    select_sst_id_set.insert(other.sst_id);
-                    extra_overlap_sst.push(other.clone());
+                    basic_overlap_info.update(other);
+
+                    add_files_size += other.get_file_size();
+                    add_files_count += 1;
                 }
 
                 if pending_compact {
                     break;
                 }
 
-                extra_overlap_levels.push((reverse_index, extra_overlap_sst));
-                *old_overlap_range = new_overlap_range;
-            }
+                // When size / file count has exceeded the limit, we need to abandon this plan, it cannot be expanded to the last sub_level
+                if ret.total_file_count > 1
+                    && (add_files_size >= self.max_compaction_bytes
+                        || add_files_count >= self.max_file_count as usize)
+                {
+                    finish = true;
+                    break;
+                }
 
-            // check reverse overlap
-            if pending_compact {
-                // encountering a pending file means we don't need to continue processing this
-                // interval
-                break;
+                extra_overlap_levels.push((reverse_index, overlap_files_range.clone()));
             }
 
-            let add_files_count = overlap_files_range.len()
-                + extra_overlap_levels
-                    .iter()
-                    .map(|(_, files)| files.len())
-                    .sum::<usize>();
-
-            if ret
-                .sstable_infos
-                .iter()
-                .filter(|ssts| !ssts.is_empty())
-                .count()
-                > MAX_COMPACT_LEVEL_COUNT
-            {
+            if pending_compact || finish {
                 break;
             }
 
-            ret.total_file_count += add_files_count;
-            ret.total_file_size += add_files_size + current_level_size;
-            if !overlap_files_range.is_empty() {
-                ret.sstable_infos[target_index]
-                    .extend_from_slice(&target_level.table_infos[overlap_files_range.clone()]);
-            }
-            overlap_len_and_begins.push(overlap_files_range);
-            for (reverse_index, files) in extra_overlap_levels {
-                ret.sstable_infos[reverse_index].extend(files);
+            for (reverse_index, sst_range) in extra_overlap_levels {
```

Comment:
nits: it is better to use `overlap_levels` instead of `extra_overlap_levels` for the naming since there is no "extra" in the current algorithm.